README.md
# NAME

CycloneDX::Parser - Parser for CycloneDX SBOMs

# VERSION

version 0.01

# SYNOPSIS

```perl
    my $parser = CycloneDX::Parser->new( json => $file );
    # or
    my $parser = CycloneDX::Parser->new( json_string => $json_string );
    if ( $parser->is_valid ) {
        my $data = $parser->sbom_data;
    }
    else {
        my @errors = $parser->errors;
        ...
    }
```

# DESCRIPTION

This module parses CycloneDX Software Bill of Materials (SBOMs), version 1.5
JSON. It is a work in progress.

Eventually earlier versions will be supported, but for now, trying to get it
working and seeing how the design evolves. The code is written with core Perl
because with the upcoming Cyber Security Act (CRA) in the EU, [open-source
code may become a
liability](https://devops.com/the-cyber-resilience-act-threatens-the-future-of-open-source/)
for many companies. By starting to build out programmatically discoverable
Software Bill of Materials (or SBOMs), we can make it easier for companies to
comply with the CRA.

Non-compliance with CRA can mean fines for companies of €15 million, or 2.5%
of global revenue, whichever is higher.  This will give many companies a
_major_ incentive to avoid using open-source, since much of it is not
designed to be easily audited.

# METHODS

## new

```perl
    my $parser = CycloneDX::Parser->new( json => $file );
    # or
    my $parser = CycloneDX::Parser->new( json_string => $json_string );
```

Creates a new parser object. The only argument is a hashref with a single key.
Thay key is `json` and the value is the JSON file to parse. If you wish to
pass in raw JSON instead of a file, use the `json_string` key instead.

## is\_valid

```
    if ( $parser->is_valid ) {
        ...
    }
```

Returns true if the SBOM is valid (warnings are OK, but errors means it's invalid).

## errors

```perl
    my @errors = $parser->errors;
```

Returns a list of errors as printable strings. If the SBOM is valid, the list
will be empty.

## warnings

```perl
    my @warnings = $parser->warnings;
```

Returns a list of warnings as printable strings. If there are no warnings, the
list will be empty.

## sbom\_data

```perl
    my $data = $parser->sbom_data;
```

Returns a hashref of the SBOM data _as\_is_. Note that this is mutable, so changing the data
means you should call `$self->validate` again.

# AUTHOR

Curtis "Ovid" Poe <curtis.poe@gmail.com>

# COPYRIGHT AND LICENSE

This software is Copyright (c) 2023 by Curtis "Ovid" Poe.

This is free software, licensed under:

```
The Artistic License 2.0 (GPL Compatible)
```


Changes
Revision history for CycloneDX-Parser

{{$NEXT}}

          - Create shell for CycloneDX::Parser


lib/CycloneDX/Parser/Checks.pm
package CycloneDX::Parser::Checks;

# ABSTRACT: Simple "type" checks for the JSON structure

use v5.14.0;
use warnings;
use experimental 'signatures';
use Carp 'croak';
use Digest::Sha 'sha1_hex', 'sha256_hex', 'sha384_hex', 'sha512_hex';

use parent 'Exporter';
our @EXPORT_OK = qw(
  any_string
  is_array_of
  is_one_of
  is_object
  is_string
  non_empty_string
);
our %EXPORT_TAGS = ( all => \@EXPORT_OK );

our $VERSION = '0.01';

# Because this is core, we're using exported subs as "poor man roles"
#
# Guidelines: C<croak()> if the error in our add code. Use C<_add_error> if
# the error is in their SBOM data.

=head2 C<is_string($thing)>

Returns a sub that will check that the value is a string. If C<$thing> is a
not a reference, it will check that the value is equal to C<$thing>. If
C<$thing> is a regular expression, it will check that the value matches the
regular expression. If C<$thing> is an array reference, it will check that the
value is one of the elements of the array.

Will croak if <$matching> is not a string, regular expression, or array.

=cut

sub is_string ($matching) {
    if ( 'ARRAY' eq ref $matching ) {

        # rewrite it as a hash for faster lookup
        $matching = { map { $_ => 1 } @$matching };
    }
    return sub ( $parser, $value ) {

        # if errors are reported, this curious little construct will make sure
        # that the error is reported with the correct sub name, not "ANON"
        local *__ANON__ = 'is_string';

        my $name = $parser->_stack;
        if ( my $ref = ref $value ) {
            croak("Value $name must be a string, not a $ref'");
        }

        if ( !ref $matching ) {
            if ( $value ne $matching ) {
                $parser->_add_error("Invalid $name. Must be '$matching', not '$value'");
                return;
            }
        }
        elsif ( ref $matching eq 'Regexp' ) {
            if ( $value !~ $matching ) {
                $parser->_add_error("Invalid $name. Must match '$matching', not '$value'");
                return;
            }
        }
        elsif ( ref $matching eq 'HASH' ) {
            if ( !$matching->{$value} ) {
                my @matching = sort keys %$matching;
                $parser->_add_error("Invalid $name. Must be one of '@matching', not '$value'");
                return;
            }
        }
        else {
            croak "Invalid matching type for is_string: $matching";
        }
        return 1;
    };
}

sub is_object ( $matching, $required = [] ) {
    if ( 'HASH' ne ref $matching ) {
        croak("is_object must be passed a hashref");
    }

    return sub ( $parser, $value ) {

        # if errors are reported, this curious little construct will make sure
        # that the error is reported with the correct sub name, not "ANON"
        local *__ANON__ = 'is_object';
        my $name = $parser->_stack;

        if ( 'HASH' ne ref $value ) {
            my $ref = _get_type($value);
            $parser->_add_error("$name: Value $name must be an object, not a $ref");
            return;
        }

        $parser->_validate(
            object   => $matching,
            source   => $value,
            required => $required,
        );
    }
}

# XXX TODOO: Make sure we can at least name whick keys fail, and where
sub is_array_of ($matches) {
    if ( 'CODE' ne ref $matches ) {
        croak("is_array_of must be passed a coderef");
    }
    return sub ( $parser, $arrayref ) {
        my $name = $parser->_stack;
        if ( 'ARRAY' ne ref $arrayref ) {
            $parser->_add_error( "Value $name must be an arrayref, not a " . _get_type($arrayref) );
            return;
        }
        my $success = 1;
        THING: for my $i ( 0 .. $#$arrayref ) {
            my $item = $arrayref->[$i];
            $parser->_push_stack($i);
            if ( !$parser->$matches($item) ) {
                $success = 0;
            }
            $parser->_pop_stack;
        }
        if ( !$success ) {
            $parser->_add_error("Invalid $name. Does not match any of the specified checks");
        }
        return $success;
    }
}

sub _get_type ($value) {
    return
        defined ref $value && ref $value ? ref $value
      : defined ref $value               ? 'scalar'
      :                                    'undef';
}

=head2 C<is_one_of>

    is_one_of(
        is_object(
            {
                color => [qw/ red green blue /],
                name  => qr/^[a-z]+$/,
            }
        ),
        is_string('Cyberdyne Systems'),
    );

Returns a sub that will check that the value is one of the things passed to it. Those "things"
need to be subroutines that return true if the value is valid.

=cut

sub is_one_of (@things) {
    if ( @things < 2 ) {
        croak("is_one_of must be passed at least two things");
    }
    if ( @things != grep { 'CODE' eq ref $_ } @things ) {
        croak("is_one_of must be passed only CODE refs");
    }
    return sub ( $parser, $value ) {
        my $name    = $parser->_stack;
        my $success = 0;

        my ( @errors, @warnings );
        THING: for my $thing (@things) {
            $parser->_stash_error_state;
            if ( $parser->$thing($value) ) {
                $success = 1;
            }
            my ( $addded_errors, $added_warnings ) = $parser->_unstash_error_state;
            push @errors,   @$addded_errors;
            push @warnings, @$added_warnings;
            last THING if $success;
        }
        if ( !$success ) {
            foreach my $error (@errors) {
                $parser->_add_error($error);
            }
            foreach my $warning (@warnings) {
                $parser->_add_warning($warning);
            }
            $parser->_add_error("Invalid $name. Does not match any of the specified checks");
        }
        return $success;
    }
}

=head2 C<any_string>

Returns a sub that will check that the value is a string. Contents of the
string do not matter.

=cut

sub any_string () {
    return is_string(qr/./);
}

=head2 C<non_empty_string>

Returns a sub that will check that the value is a string and that it contains
at least one non-whitespace character.

=cut

sub non_empty_string () {
    return is_string(qr/\S/);
}

1;


lib/CycloneDX/Parser/v1_5/JSON.pm
package CycloneDX::Parser::v1_5::JSON;

# ABSTRACT: JSON parser for CycloneDX SBOMs, v1.5

use v5.14.0;
use warnings;
use experimental 'signatures';

use Carp qw(cluck carp croak);
use JSON::PP 'decode_json';

use CycloneDX::Parser::Checks ':all';
use parent 'Exporter';

our @EXPORT_OK = qw(
  sbom_spec
);
our %EXPORT_TAGS = ( all => \@EXPORT_OK );

our $VERSION = '0.01';

sub sbom_spec () {
    return {
        object => {
            bomFormat    => is_string('CycloneDX'),
            specVersion  => is_string('1.5'),
            components   => \&_validate_components,
            serialNumber => is_string(qr/^urn:uuid:[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/),
            version      => is_string(qr/^[1-9][0-9]*$/),
            metadata     => is_object(
                {
                    timestamp  => is_string(qr/^\d{4}-\d\d-\d\dT\d\d:\d\d:\d\d(\+\d\d:\d\d)?Z?$/),
                    properties => is_array_of(
                        is_object(
                            {
                                name  => non_empty_string,    # valid-properties not failing if I put an empty name
                                value => non_empty_string,
                            }
                        ),
                    ),
                    lifecycles => is_array_of(
                        is_one_of(
                            is_object(
                                {
                                    phase => is_string( [ "design", "pre-build", "build", "post-build", "operations", "discovery", "decommission" ] ),
                                },
                                ['phase']
                            ),
                            is_object( { name => any_string, description => any_string }, ['name'] ),
                        )
                    ),
                },
            ),

            # lifecycles
            # services
            # dependencies
            # externalReferences
            # properties
            # vulnerabilities
            # annotations
            # formulation
            # properties
            # signature
        },
        # for 1.5, these are the only required fields. `version` is no longer required
        # because if it's missing, it has an optional value of 1.
        required => [qw(bomFormat specVersion )],
    };
}

sub _validate_components ( $self, $components ) {
    unless ( ref $components eq 'ARRAY' ) {
        $self->_add_error('Components must be an array');
        return;
    }

    foreach my $i ( 0 .. $#$components ) {
        $self->_push_stack($i);
        my $component = $components->[$i];
        $self->_validate(
            object => {
                name => is_string(qr/\S/),
                type => is_string(
                    [   "application",
                        "framework",
                        "library",
                        "container",
                        "platform",
                        "operating-system",
                        "device",
                        "device-driver",
                        "firmware",
                        "file",
                        "machine-learning-model",
                        "data",
                    ],
                ),
                version     => non_empty_string,                                # version not enforced
                'mime-type' => is_string( qr{^[-+a-z0-9.]+/[-+a-z0-9.]+$}, ),
                'bom-ref'   => non_empty_string,
                author      => any_string,
                publisher   => any_string,
                group       => any_string,
                description => any_string,
                scope       => is_string( [qw/required optional excluded/] ),
                copyright   => any_string,
                cpe         => any_string,                                      # dont' have great info on matching a "well-formed CPE string"
                    # See # https://metacpan.org/dist/URI-PackageURL/source/lib/URI/PackageURL.pm # for purl
                purl       => is_string(qr{^pkg:[A-Za-z\\.\\-\\+][A-Za-z0-9\\.\\-\\+]*/.+}),
                components => \&_validate_components,                                          # yup, components can take components

                # supplier is an object
                # hashes is an array of objects
                # licenses is an array of objects
                # externalReferences is an array of objects
                # data is an array of objects
                # properties is an array of objects
                # swid is an object
                # pedigree is an object
                # evidence is an object
                # releaseNotes is an object
                # modelCard is an object
                # signature is an object
            },
            required => [qw( name type )],
            source   => $component,
        );

        my $name = $self->_stack;

        # bom-ref must be unique
        if ( exists $component->{'bom-ref'} ) {
            if ( $self->_bom_ref_seen( $component->{'bom-ref'} ) ) {
                $self->_add_error( sprintf "$name.bom-ref: Duplicate bom-ref '%s'", $component->{'bom-ref'} );
            }

            # XXX later, we'll have more validation
        }

        if ( exists $component->{modified} ) {
            $self->_add_warning('$name.modified is deprecated and should not be used.');
        }

        $self->_pop_stack;
    }
}

1;


lib/CycloneDX/Parser.pm
package CycloneDX::Parser;

# ABSTRACT: Parser for CycloneDX SBOMs

use v5.14.0;
use warnings;
use experimental 'signatures';

use Carp qw(carp croak);
use JSON::PP 'decode_json';
use Module::Load 'load';

our $VERSION = '0.01';

sub new ( $class, %arg_for ) {
    my $self = bless {}, $class;
    $self->_initialize(%arg_for);
    return $self;
}

sub _initialize ( $self, %arg_for ) {
    my $filename    = $arg_for{json} // '';
    my $json_string = $arg_for{json_string};

    if ( $filename && $json_string ) {
        my $class = ref $self;
        croak "You must specify only one of 'json' and 'json_string' when contructing a $class";
    }

    if ($filename) {
        open my $fh, '<', $filename or croak "Can't open $filename for reading: $!";
        $json_string = do { local $/; <$fh> };
        close $fh or croak "Can't close $filename: $!";
    }

    $self->{sbom_data} = eval {
        decode_json($json_string);                 # the JSON as a Perl structure
    } or do {
        croak "Invalid JSON in $filename: $@";
    };
    my $specVersion = $self->sbom_data->{specVersion} // croak("No specVersion specified");

    $self->{specVersion} = $specVersion;           # CycloneDX spec version
    $self->{filename}    = $filename;              # the source of the JSON, if file passed
    $self->{json}        = $json_string;           # the JSON as a string
    $self->{debug}       = $arg_for{debug} // 0;
    $self->{error_state} = {                       # will be used to stash error state from recoverable errors
        stashed  => undef,                         # have we previously stashed an error?
        errors   => [],
        warnings => [],
    };

    $self->_load_version($specVersion);
    $self->validate;

    if ( $self->has_warnings ) {

        # note: currenty the only warnings is about the deprected 'modified' field
        warn "Warnings:\n";
        foreach my $warning ( $self->warnings ) {
            warn "  $warning\n";
        }
    }

    return $self;
}

sub _load_version($self, $specVersion) {
    my $version = "v$specVersion";
    $version =~ s/\./_/g;
    #use CycloneDX::Parser::v1_5::JSON;
    my $module = "CycloneDX::Parser::${version}::JSON"; # later, maybe XML
    load $module, qw(sbom_spec);
}

sub validate ($self) {

    # make sure theyse are empty before we start validation.
    $self->{errors}        = [];    # accumulate errors
    $self->{warnings}      = [];    # accumulate warnings
    $self->{stack}         = [];    # track the current location in the JSON
    $self->{bom_refs_seen} = {};    # track bom-ref ids to ensure they are unique

    my $spec = sbom_spec();
    $self->_validate(
        object   => $spec->{object},
        required => $spec->{required},
        source   => $self->sbom_data,
    );
}

sub json ($self) {
    return $self->{json};
}

sub is_valid ($self) {
    return !@{ $self->{errors} };
}

sub errors ($self) {
    return @{ $self->{errors} };
}

sub warnings ($self) {
    return @{ $self->{warnings} };
}

sub has_warnings ($self) {
    return !!@{ $self->{warnings} };
}

sub sbom_data ($self) {
    return $self->{sbom_data};
}

sub set_debug_level ( $self, $level ) {
    $self->{debug} = $level;
}

sub _stash_error_state ($self) {
    my $error_state = $self->{error_state};
    my $stack       = $self->_stack;
    $self->_debug( 1, "Stashing error state at $stack" );
    $self->_debug_dump( 1, { error_state => $error_state, errors => $self->{errors}, warnings => $self->{warnings} } );
    if ( my $stack = $error_state->{stashed} ) {

        # I might want a true stack here and push errors onto
        # the stack and pop them off as appropriate
        croak "Can't stash error state twice: $stack";
    }
    $error_state->{stashed} = $self->_stack;

    # shallow copies to avoid referencing
    $error_state->{errors}   = [ @{ $self->{errors} } ];
    $error_state->{warnings} = [ @{ $self->{warnings} } ];
}

sub _unstash_error_state ($self) {
    my $error_state = $self->{error_state};
    if ( !$error_state->{stashed} ) {
        croak "Can't unstash error state when it hasn't been stashed";
    }

    my $stashed_error_state   = $error_state->{errors};
    my $stashed_warning_state = $error_state->{warnings};
    my $old_error_count       = @$stashed_error_state;
    my $old_warning_count     = @$stashed_warning_state;

    my @new_errors   = @{ $self->{errors} };
    my @new_warnings = @{ $self->{warnings} };

    my @added_errors   = splice @new_errors,   0, $old_error_count;
    my @added_warnings = splice @new_warnings, 0, $old_warning_count;

    # shallow copies to avoid referencing
    $self->{errors}   = [@$stashed_error_state];
    $self->{warnings} = [@$stashed_warning_state];

    $error_state->{stashed} = undef;
    my $stack = $self->_stack;
    $self->_debug( 0, "Unstashed error state at $stack" );
    $self->_debug_dump( 0, { error_state => $error_state, errors => $self->{errors}, warnings => $self->{warnings} } );

    return \@added_errors, \@added_warnings;
}

sub _add_error ( $self, $error ) {
    push @{ $self->{errors} }, $error;
}

sub _add_warning ( $self, $warning ) {
    push @{ $self->{warnings} }, $warning;
}

sub _push_stack ( $self, $name ) {
    if ( $self->_is_debugging ) {
        my $stack = $self->_stack;
        $self->_debug( 1, "Pushing $name onto stack: $stack" );
    }
    push @{ $self->{stack} }, $name;
}

sub _pop_stack ($self) {
    my $name = pop @{ $self->{stack} };
    if ( $self->_is_debugging ) {
        my $stack = $self->_stack;
        $self->_debug( 0, "Popped $name off stack: $stack" );
    }
}

sub _is_debugging ($self) {
    return $self->{debug};
}

sub _debug_leader ( $self, $char ) {
    my $stack = $self->_stack;
    my $num   = 1 + $stack =~ tr/././;
    $num *= 2;
    return $char x $num;
}

sub _debug ( $self, $in, $string ) {
    return unless $self->_is_debugging;
    my $leader = $self->_debug_leader( $in ? '>' : '<' );
    $string = "$leader $string";
    1 == $self->{debug} ? carp $string : carp $string;
}

sub _debug_dump ( $self, $in, $data ) {
    return unless $self->_is_debugging > 1;
    my $leader = $self->_debug_leader( $in ? '>' : '<' );
    my $spaces = ' ' x length $leader;

    require Data::Dumper;
    no warnings 'once';    ## no critic (TestingAndDebugging::ProhibitNoWarnings)
    local $Data::Dumper::Terse     = 1;
    local $Data::Dumper::Indent    = 1;
    local $Data::Dumper::Sortkeys  = 1;
    local $Data::Dumper::Quotekeys = 0;
    $data = Data::Dumper::Dumper($data);
    $data =~ s/^/$spaces /mg;
    $data =~ s/$spaces/$leader/;
    carp $data;
}

sub _stack ($self) {
    my $stack = join '.', @{ $self->{stack} };
    $stack = "<unknown>" if !length $stack;
    return $stack;
}

# Yes, this should use a JSON validator, but we need it to be lightweight
# and not require any non-core modules.
# Also, "No additional properties" is hard to properly handle until we have
# the full list of properties included.
sub _validate ( $self, %arg_for ) {
    my $required = $arg_for{required} // [];
    my $source   = $arg_for{source};

    my %is_required = map { $_ => 1 } @$required;

    my @errors = $self->errors;
    KEY: foreach my $key ( sort keys %{ $arg_for{object} } ) {
        my $matches = $arg_for{object}{$key};
        $self->_push_stack($key);
        my $name = $self->_stack;
        if ( !exists $source->{$key} ) {
            if ( $is_required{$key} ) {
                $self->_add_error("Missing required field '$name'");
            }
        }
        else {
            my $value = $source->{$key};
            if ( ref $matches eq 'CODE' ) {
                $self->$matches($value);
            }
            else {
                croak "Invalid matches type: $matches";
            }
        }
        $self->_pop_stack;
    }
    if ( $self->errors > @errors ) {

        # new errors were found, so we return false
        return;
    }
    return 1;
}

sub _bom_ref_seen ( $self, $bom_ref ) {
    my $seen = $self->{bom_refs_seen}{$bom_ref};
    $self->{bom_refs_seen}{$bom_ref}++;
    return $seen;
}

1;

__END__

=head1 SYNOPSIS

        my $parser = CycloneDX::Parser->new( json => $file );
        # or
        my $parser = CycloneDX::Parser->new( json_string => $json_string );
        if ( $parser->is_valid ) {
            my $data = $parser->sbom_data;
        }
        else {
            my @errors = $parser->errors;
            ...
        }


=head1 DESCRIPTION

This module parses CycloneDX Software Bill of Materials (SBOMs), version 1.5
JSON. It is a work in progress.

Eventually earlier versions will be supported, but for now, trying to get it
working and seeing how the design evolves. The code is written with core Perl
because with the upcoming Cyber Security Act (CRA) in the EU, L<open-source
code may become a
liability|https://devops.com/the-cyber-resilience-act-threatens-the-future-of-open-source/>
for many companies. By starting to build out programmatically discoverable
Software Bill of Materials (or SBOMs), we can make it easier for companies to
comply with the CRA.

Non-compliance with CRA can mean fines for companies of €15 million, or 2.5%
of global revenue, whichever is higher.  This will give many companies a
I<major> incentive to avoid using open-source, since much of it is not
designed to be easily audited.

=head1 METHODS

=head2 new

        my $parser = CycloneDX::Parser->new( json => $file );
        # or
        my $parser = CycloneDX::Parser->new( json_string => $json_string );

Creates a new parser object. The only argument is a hashref with a single key.
Thay key is C<json> and the value is the JSON file to parse. If you wish to
pass in raw JSON instead of a file, use the C<json_string> key instead.

=head2 is_valid

        if ( $parser->is_valid ) {
            ...
        }

Returns true if the SBOM is valid (warnings are OK, but errors means it's invalid).

=head2 errors

        my @errors = $parser->errors;

Returns a list of errors as printable strings. If the SBOM is valid, the list
will be empty.

=head2 warnings

        my @warnings = $parser->warnings;

Returns a list of warnings as printable strings. If there are no warnings, the
list will be empty.

=head2 sbom_data

        my $data = $parser->sbom_data;

Returns a hashref of the SBOM data I<as_is>. Note that this is mutable, so changing the data
means you should call C<< $self->validate >> again.
